### PR TITLE
Create Operand Form for OLM

### PR DESCRIPTION
**Pull request by alecmerdler**
### Description

Initial implementation of an alternative to the YAML editor creation flow, powered by the [Operator Lifecycle Manager's `specDescriptors`](https://github.com/openshift/console/tree/master/frontend/public/components/operator-lifecycle-manager/descriptors) and [Kubernetes CustomResourceDefinition's JSONSchema](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#validation). 

**Try it:** `quay.io/alecmerdler/console:OLM-695`

### Screenshots

**Create Operand Form:**
Screenshot_20190619_154024

**Create Operand Form (error):**
Screenshot_20190620_134617

**Create Operand YAML:**
Screenshot_20190619_154035

### Future Work

- Implement more descriptor components
- Input validation using JSONSchema from CRD (if present)
- Support grouping inputs
- Maintain changes between form/YAML editor
- Document how to contribute new form input components

Addresses https://jira.coreos.com/browse/OLM-695

### CODE REVIEW

File: frontend/public/components/edit-yaml.jsx [hunk 1]
```diff
@@ -356,7 +356,7 @@ export const EditYAML = connect(stateToProps)(
         { canDrop && <div className={klass}><p className="co-file-dropzone__drop-text">Drop file here</p></div> }
 
         <div>
-          {create && <div className="yaml-editor__header">
+          {create && this.props.showHeader && <div className="yaml-editor__header">
```
Comment (alecmerdler) [inline frontend/public/components/edit-yaml.jsx]: Introduced this so we can show the same header with breadcrumbs between form/YAML editor.

File: frontend/public/components/modals/configure-count-modal.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import * as _ from 'lodash-es';
+import * as React from 'react';
+
+import { k8sPatch, K8sResourceKind, K8sKind } from '../../module/k8s';
+import { createModalLauncher, ModalTitle, ModalBody, ModalSubmitFooter } from '../factory/modal';
+import { NumberSpinner, withHandlePromise } from '../utils';
+
+export const ConfigureCountModal = withHandlePromise((props: ConfigureCountModalProps) => {
```
Comment (alecmerdler) [inline frontend/public/components/modals/configure-count-modal.tsx]: Refeactor to function component.

File: frontend/public/components/operator-lifecycle-manager/create-operand.tsx [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+      },
+      spec: _.reduce(_.omitBy(formValues, (val, path) => path.startsWith('metadata')), (spec, value, path) => _.set(spec, path, value), {}),
+    } as K8sResourceKind;
+
+    k8sCreate(props.operandModel, obj)
+      .then(() => history.push(`/k8s/ns/${props.namespace}/clusterserviceversions/${props.clusterServiceVersion.metadata.name}/${referenceForModel(props.operandModel)}/${obj.metadata.name}`))
+      .catch((err: Status) => setError(err.message));
+  };
+
+  const inputFor = (field: OperandField) => {
```
Comment (alecmerdler) [inline frontend/public/components/operator-lifecycle-manager/create-operand.tsx]: More input fields will be implemented as follow-up stories.

File: frontend/public/components/operator-lifecycle-manager/create-operand.tsx [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+  }
+
+  return <CreateYAML template={!_.isError(template) ? template : null} match={props.match} showHeader={false} />;
+};
+
+export const CreateOperand: React.FC<CreateOperandProps> = (props) => {
+  const providedAPI = () => providedAPIsFor(props.clusterServiceVersion.data).find(crd => referenceForProvidedAPI(crd) === referenceForModel(props.operandModel));
+  const sample = () => JSON.parse(_.get(props.clusterServiceVersion.data.metadata.annotations, annotationKey, '[]'))
+    .find((s: K8sResourceKind) => referenceFor(s) === referenceForModel(props.operandModel));
+  const [method, setMethod] = React.useState<'yaml' | 'form'>('yaml');
```
Comment (alecmerdler) [inline frontend/public/components/operator-lifecycle-manager/create-operand.tsx]: Show YAML editor by default for now until form is more feature rich.

File: frontend/public/components/operator-lifecycle-manager/descriptors/spec/index.tsx [hunk 5]
```diff
@@ -41,35 +41,35 @@ const K8sResourceLink: React.SFC<SpecCapabilityProps> = (props) => _.isEmpty(pro
...
-      &nbsp;&nbsp;
-      {state.value !== props.value && state.confirmed && <LoadingInline />}
-      {state.value !== props.value && !state.confirmed && <React.Fragment>
-        &nbsp;&nbsp;<i className="fa fa-exclamation-triangle text-warning" aria-hidden="true" />
-        <button className="btn btn-link" onClick={update}>Confirm change</button>
-      </React.Fragment>}
-    </div>;
-  }
-}
+const BooleanSwitch: React.FC<SpecCapabilityProps> = (props) => {
```
Comment (alecmerdler) [inline frontend/public/components/operator-lifecycle-manager/descriptors/spec/index.tsx]: Refactor to function component.

File: frontend/public/components/operator-lifecycle-manager/descriptors/spec/resource-requirements.tsx [hunk 6]
```diff
@@ -8,6 +8,21 @@ import { k8sUpdate, referenceFor, K8sKind } from '../../../../module/k8s';
 import { createModalLauncher, ModalTitle, ModalBody, ModalSubmitFooter } from '../../../factory/modal';
 import { RootState } from '../../../../redux';
 
+export const ResourceRequirements: React.FC<ResourceRequirementsProps> = (props) => {
```
Comment (alecmerdler) [inline frontend/public/components/operator-lifecycle-manager/descriptors/spec/resource-requirements.tsx]: Split this out of the modal to use as an inline component.

**Comment by spadgett** [comment]
cc @alimobrem @jwforres

**Comment by spadgett** [comment]
In other contexts like this, the link is "Edit YAML":

OKD 2019-06-24 13-13-11

> Maintain changes between form/YAML editor

There are some designs that make the form and YAML tabs that you can switch back and forth between.

File: frontend/public/components/create-yaml.tsx [hunk 7]
```diff
@@ -11,7 +11,7 @@ import { ErrorPage404 } from './error';
 import { ClusterServiceVersionModel } from '../models';
 
 export const CreateYAML = connectToPlural((props: CreateYAMLProps) => {
-  const {match, kindsInFlight, kindObj} = props;
+  const {match, kindsInFlight, kindObj, showHeader = true} = props;
```
Comment (spadgett) [inline frontend/public/components/create-yaml.tsx]: Booleans that default to true are confusing.  I suggest `hideHeader` and default to false

File: frontend/public/components/operator-lifecycle-manager/_operator-lifecycle-manager.scss [hunk 8]
```diff
@@ -140,3 +140,22 @@
...
+
+.co-create-operand__header {
+  border-bottom: 1px solid #ddd;
+  padding: 10px 15px;
+  @media(min-width: $screen-sm-min) {
+    padding: 10px 31px; // left alignment with namespace selector text
+  }
+}
+
+.co-create-operand__header__buttons {
```
Comment (spadgett) [inline frontend/public/components/operator-lifecycle-manager/_operator-lifecycle-manager.scss]: Avoid nesting blocks in BEM. See http://getbem.com/faq/

Maybe...?

```suggestion
.co-create-operand__header-buttons {
```

File: frontend/public/components/operator-lifecycle-manager/create-operand.tsx [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+  return <React.Fragment>
+    { props.loaded && <div className="co-create-operand__header">
+      <div className="co-create-operand__header__buttons">
+        <BreadCrumbs breadcrumbs={[
+          {name: props.clusterServiceVersion.data.spec.displayName, path: window.location.pathname.replace('/~new', '')},
+          {name: `Create ${props.operandModel.label}`, path: window.location.pathname},
+        ]} />
+        <div style={{marginLeft: 'auto'}}>{
+          method === 'form' && <button className="btn btn-link" onClick={() => setMethod('yaml')}>YAML Editor</button> ||
```
Comment (spadgett) [inline frontend/public/components/operator-lifecycle-manager/create-operand.tsx]: We might eventually make these tabs.

```suggestion
          method === 'form' && <button className="btn btn-link" onClick={() => setMethod('yaml')}>Edit YAML</button> ||
```

File: frontend/public/components/operator-lifecycle-manager/create-operand.tsx [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+          method === 'yaml' && <button className="btn btn-link" onClick={() => setMethod('form')}>Form</button>
```
Comment (spadgett) [inline frontend/public/components/operator-lifecycle-manager/create-operand.tsx]:
```suggestion
          method === 'yaml' && <button className="btn btn-link" onClick={() => setMethod('form')}>Edit Form</button>
```

File: frontend/public/components/operator-lifecycle-manager/create-operand.tsx (hunk 9 above)
Comment (alecmerdler) [reply frontend/public/components/operator-lifecycle-manager/create-operand.tsx]: I tried tabs at first, but it was pretty hard to wrangle the existing `HorizontalNav` components to perform the right way. Also the existing pattern is that tabs show different info, not the same info in different ways, which I thought was confusing to use.

**Comment by alecmerdler** [comment]
Addressed comments and added more unit tests.

**Pull request merged**